src/XR/features/WebXRNearInteraction.ts

```diff
@@ -0,0 +1,590 @@
+import { WebXRControllerPointerSelection } from "./WebXRControllerPointerSelection";
+
+/**
+ * Options interface for the near interaction pointer selection module
+ */
+export interface IWebXRNearInteractionOptions {
+    /**
+     * if provided, this scene will be used to render meshes.
+     */
+    customUtilityLayerScene?: Scene;
+    /**
+     * Should meshes created here be added to a utility layer or the main scene
+     */
+    useUtilityLayer?: boolean;
+    /**
+     * the xr input to use with this pointer selection
+     */
+    xrInput: WebXRInput;
+
+    farInteractionFeature: WebXRControllerPointerSelection;
+    /**
+     * Enable pointer selection on all controllers instead of switching between them
+     */
+    enablePointerSelectionOnAllControllers?: boolean;
+    /**
+     * The preferred hand to give the pointer selection to. This will be prioritized when the controller initialize.
+     * If switch is enabled, it will still allow the user to switch between the different controllers
+     */
+    preferredHandedness?: XRHandedness;
+    /**
+     * Disable switching the pointer selection from one controller to the other.
+     * If the preferred hand is set it will be fixed on this hand, and if not it will be fixed on the first controller added to the scene
+     */
+    disableSwitchOnClick?: boolean;
+}
+
+/**
+ * A module that will enable near interaction pointer selection for hands and motion controllers of XR Input Sources
+ */
+export class WebXRNearInteraction extends WebXRAbstractFeature {
+    private static _idCounter = 200;
+
+    private _attachController = (xrController: WebXRInputSource) => {
+        if (this._controllers[xrController.uniqueId]) {
+            // already attached
+            return;
+        }
+        const { hoverIndexMeshTip, pickIndexMeshTip } = this._generateNewHandTipMeshes();
+
+        // get two new meshes
+        this._controllers[xrController.uniqueId] = {
+            xrController,
+            meshUnderPointer: null,
+            pick: null,
+            pickIndexMeshTip,
+            hoverIndexMeshTip,
+            grabRay: new Ray(new Vector3(), new Vector3()),
+            nearInteraction: false,
+            nearGrab: false,
+            nearGrabInProcess: false,
+            id: WebXRNearInteraction._idCounter++,
+        };
+
+        if (this._attachedController) {
+            if (
+                !this._options.enablePointerSelectionOnAllControllers &&
+                this._options.preferredHandedness &&
+                xrController.inputSource.handedness === this._options.preferredHandedness
+            ) {
+                this._attachedController = xrController.uniqueId;
+            }
+        } else {
+            if (!this._options.enablePointerSelectionOnAllControllers) {
+                this._attachedController = xrController.uniqueId;
+            }
+        }
+        // NOT SURE: if i should return null for gaze and screen modes

```

Comment:
I don't think it makes sense to perform any near interaction logic using the gaze/screen, so since this is now a standalone class, you can return even earlier in those cases, maybe before even calling into _attachController().